modules/core/image/src/Plugin/ImageEffect/AutoOrientImageEffect.php

```diff
@@ -0,0 +1,217 @@
+    // This logic is adapted from the Drupal Image Effects module.
+    // @see https://git.drupalcode.org/project/image_effects/-/blob/730a1e6f5f947ef4d6eec5d359c6968ce93594f9/src/Plugin/ImageToolkit/Operation/gd/AutoOrient.php#L45-86
+    // http://www.sno.phy.queensu.ca/~phil/exiftool/TagNames/EXIF.html:
+    // 1 = Horizontal (normal)                 [top-left].
+    // 2 = Mirror horizontal                   [top-right].
+    // 3 = Rotate 180                          [bottom-right].
+    // 4 = Mirror vertical                     [bottom-left].
+    // 5 = Mirror horizontal and rotate 270 CW [left-top].
+    // 6 = Rotate 90 CW                        [right-top].
+    // 7 = Mirror horizontal and rotate 90 CW  [right-bottom].
+    // 8 = Rotate 270 CW                       [left-bottom].
+    switch ($orientation) {
+      case 2:
+        $result = $image->apply('mirror', ['x_axis' => TRUE]);
+        break;
+
+      case 3:
+        $result = $image->apply('rotate', ['degrees' => 180]);
+        break;
+
+      case 4:
+        $result = $image->apply('mirror', ['y_axis' => TRUE]);
+        break;
+
+      case 5:
+        $result = $image->apply('mirror', ['x_axis' => TRUE]);
+        if ($result) {
+          $result = $image->apply('rotate', ['degrees' => 270]);
+        }
+        break;
+
+      case 6:
+        $result = $image->apply('rotate', ['degrees' => 90]);
+        break;
+
+      case 7:
+        $result = $image->apply('mirror', ['x_axis' => TRUE]);
+        if ($result) {
+          $result = $image->apply('rotate', ['degrees' => 90]);
+        }
+        break;
+
+      case 8:
+        $result = $image->apply('rotate', ['degrees' => 270]);
+        break;
+
+      default:
+        $result = TRUE;
+    }
+
+    // If the transformation failed, bail.
+    if (!$result) {
+      $this->logger->error('Image auto-orient failed using the %toolkit toolkit on %path (%mimetype)', [
+        '%toolkit' => $image->getToolkitId(),
+        '%path' => $image->getSource(),
+        '%mimetype' => $image->getMimeType(),
+      ]);
+      return $result;
+    }
+
+    // Return the result.
+    return $result;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function transformDimensions(array &$dimensions, $uri) {
+
+    // Only proceed if the exif PHP extension is loaded.
+    if (!extension_loaded('exif')) {
+      return;
+    }
+
+    // This logic is adapted from the Drupal Image Effects module.
+    // @see https://git.drupalcode.org/project/image_effects/-/blob/730a1e6f5f947ef4d6eec5d359c6968ce93594f9/src/Plugin/ImageEffect/AutoOrientImageEffect.php
+    $dimensions['width'] = $dimensions['width'] ? (int) $dimensions['width'] : NULL;
+    $dimensions['height'] = $dimensions['height'] ? (int) $dimensions['height'] : NULL;
+    if ($dimensions['width'] && $dimensions['height']) {
+      $orientation = $this->loadExifOrientation($uri);
+      if (in_array($orientation, [5, 6, 7, 8])) {
```

Comment:
Check that returned orientation value is not null

```suggestion
      $orientation = $this->loadExifOrientation($uri);
      if ($orientation && in_array($orientation, [5, 6, 7, 8])) {
```

Reply:
I had that originally and PHPStorm highlighted it, saying:

> Condition is unnecessary because it is checked by 'in_array($orientation, [5, 6, 7, 8])' 

In other words: `NULL` is not in the array `[5, 6, 7, 8]`, so it will work as expected.

Reply:
Oh, great!